Zstdsnapshoting compression

willyborankin opened the pull request:
### Description
This PR adds support ZSTD compression for snapshoting metadata. ZSTD compression for indexes is out of scope since such compression must be supported by Lucene. 
There is a PR in Lucene https://github.com/apache/lucene/pull/439 with support of ZSTD, but they do not want to merge it so far.     
  

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/2192
 
### Check List
- [] New functionality includes testing.
  - [] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock reviewed (commented):
A cursory look at this looks good to me. 

How does one use it?

I'd like @nknize to take a look please.

willyborankin commented:
> A cursory look at this looks good to me.
> 
> How does one use it?
> 
> I'd like @nknize to take a look please.

During creation of a repository set compression type (default is deflate, how it works now):
```json
{
  ...
   "settings": {
      "compress": true,
      "compression_type": "zstd", // `deflate`, `lz4`
   }
}
```

reta commented on server/src/main/resources/org/opensearch/bootstrap/security.policy:
```diff
@@ -182,4 +182,6 @@ grant {
   permission java.io.FilePermission "/sys/fs/cgroup/memory/-", "read";
 
+  // ZSTD permissions
+  permission java.lang.RuntimePermission "*";
```
:( Please fix that to have narrowed set of permissions

nknize replied:
+1; we can't have the security here be wide open

willyborankin replied:
Yes it is a bit complex thing. What would be better:
- change it to `java.lang.RuntimePermission "loadLibrary.*"`
- Alternative solution is to re-pack zstd library and add *.so files in the `libs` folder?

nknize commented on server/src/main/java/org/opensearch/common/compress/NullCompressor.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @opensearch.internal
+ */
+public class NullCompressor implements Compressor {
```
`CompressorFactory.compressor` is nullable, why do we need an explicit null compressor? Can we just check `compressor == null ? indexOutputOutputStream : compressor.threadLocalOutputStream(indexOutputOutputStream)` and do away with this empty class?

willyborankin replied:
TBH it is just a habit to use Null class and do not check it on null every where, but I think with None it would be better

nknize commented on server/src/main/java/org/opensearch/common/compress/ZstdCompressor.java:
```diff
@@ -0,0 +1,101 @@
+    // It needs to be different from other compressors and to not be specific
+    // enough so that no stream starting with these bytes could be detected as
+    // a XContent
+    private static final byte[] HEADER = new byte[] { 'Z', 'S', 'T', 'D', '\0' };
```
readability nit: 
```suggestion
    public static final String NAME = "ZSTD";
    private static final byte[] HEADER =NAME.getBytes();
```

willyborankin replied:
Well I just wanted to do it as it done for `DEFLATE { 'D', 'F', 'L', '\0' } ` not sure about `\0`

nknize commented on test/framework/src/main/java/org/opensearch/repositories/blobstore/OpenSearchBlobStoreRepositoryIntegTestCase.java:
```diff
@@ -92,7 +93,13 @@ public static RepositoryData getRepositoryData(Repository repository) {
+        final boolean compress = randomBoolean();
+        settingsBuilder.put("compress", compress);
+        if (compress) {
+            settingsBuilder.put("compression_type", randomFrom(CompressorType.values()));
```
This doesn't test no compression or LZ4 since `CompressorType` doesn't contain those values.

nknize commented on test/framework/src/main/java/org/opensearch/snapshots/AbstractSnapshotIntegTestCase.java:
```diff
@@ -352,7 +353,11 @@ protected void createRepository(String repoName, String type) {
+        final boolean compress = randomBoolean();
+        settings.put("location", randomRepoPath()).put("compress", compress);
+        if (compress) {
+            settings.put("compression_type", randomFrom(CompressorType.values()));
```
Same here: LZ4 testing is removed and no compression testing isn't covered.

willyborankin replied:
Same here I'm planning to add LZ4 as a separate PR if you do not mind.

willyborankin commented:
> Looking pretty good! Thanks for doing this. I left some comments.
> 
> I think we should add a `none` `compression_type` option to the DSL API that nulls out the compressor instead of adding a NullCompressor:
> 
> ```
> {
>   ...
>    "settings": {
>       "compress": true,
>       "compression_type": "zstd", // `deflate`, `lz4`, `none`
>    }
> }
> ```
> 
> We also need to widen the test coverage to include LZ4 and no compression.

Yes will `none` much better than Null will change

Poojita-Raj commented:
@willyborankin Hey, would you like to address the gradle check failure and merge this PR?

wbeckler commented:
@willyborankin Have you ever benchmarked this to measure the actual impact on snapshotting and restore?

willyborankin commented:
Hi @scrawfor99, fixed diff. AFAIR I addressed all changes @nknize asked to do a long time ago. After that I was waiting for the plugin for index compression to be merged. I fixed all conflicts.

reta commented on server/src/main/java/org/opensearch/common/compress/NoneCompressor.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public BytesReference uncompress(BytesReference bytesReference) throws IOException {
+        return null;
```
```suggestion
        return bytesReference;
```

reta commented on server/src/test/java/org/opensearch/common/compress/AbstractCompressorTests.java:
```diff
@@ -0,0 +1,408 @@
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+
+public abstract class AbstractCompressorTests extends OpenSearchTestCase {
```
Probably no need to make this base test `public`

willyborankin replied:
I'm sorry do not get it. It is already public

reta commented on server/src/main/java/org/opensearch/common/compress/ZstdCompressor.java:
```diff
@@ -0,0 +1,101 @@
+        if (Arrays.equals(header, HEADER) == false) {
+            throw new IllegalArgumentException("Input stream is not compressed with ZSTD!");
+        }
+        return new BufferedInputStream(new ZstdInputStreamNoFinalizer(in, RecyclingBufferPool.INSTANCE), BUFFER_SIZE);
```
So I am unsure we need to wrap `ZstdInputStreamNoFinalizer` into `BufferedInputStream` ? (because it is already using buffers internally), may be we need 

```
new ZstdInputStreamNoFinalizer(new BufferedInputStream(in, BUFFER_SIZE), RecyclingBufferPool.INSTANCE));
```

? (also suggested here https://github.com/luben/zstd-jni/issues/194#issuecomment-974647924)

willyborankin replied:
Yeap. I will change it

reta commented on server/src/main/java/org/opensearch/common/compress/CompressorFactory.java:
```diff
@@ -61,6 +68,9 @@ public static Compressor compressor(BytesReference bytes) {
             // as a xcontent, we have a problem
             assert XContentHelper.xContentType(bytes) == null;
             return COMPRESSOR;
```
I think you should be returning `DEFLATE_COMPRESSOR` here (and checking it above), not `COMPRESSOR`

willyborankin replied:
It is ` DEFLATE_COMPRESSOR`. I marked `COMPRESSOR` as deprecated and assigned it to `DEFLATE_COMPRESSOR`. The reason is that `COMPRESSOR`  in code

reta replied:
This is why it would make sense to use DEFLATE_COMPRESSOR - COMPRESSOR  should be gone

willyborankin replied:
aye aye, sir :-)

dblock commented on CHANGELOG.md:
```diff
@@ -50,6 +51,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Change http code on create index API with bad input raising NotXContentException from 500 to 400 ([#4773](https://github.com/opensearch-project/OpenSearch/pull/4773))
 - Change http code for DecommissioningFailedException from 500 to 400 ([#5283](https://github.com/opensearch-project/OpenSearch/pull/5283))
 - Improve summary error message for invalid setting updates ([#4792](https://github.com/opensearch-project/OpenSearch/pull/4792))
+- Add ZSTD compression type for snapshotting
```
Add PR link here pls.

willyborankin replied:
added

dblock reviewed (commented):
Looks great!

willyborankin commented on server/src/main/resources/org/opensearch/bootstrap/security.policy:
```diff
@@ -79,6 +79,12 @@ grant codeBase "${codebase.jna}" {
 };
 
+
+//ZSTD compression
```
fixed

reta commented on server/src/main/java/org/opensearch/common/compress/DeflateCompressor.java:
```diff
@@ -156,17 +156,10 @@ public InputStream threadLocalInputStream(InputStream in) throws IOException {
-            len += read;
-        }
-        if (len != HEADER.length || Arrays.equals(headerBytes, HEADER) == false) {
+    public InputStream inputStream(InputStream in, boolean threadLocal) throws IOException {
```
Hm ... why we removed `static` from method signature?

reta commented:
FYI, there is only one plugin (outside core) that implements repository for OCI [1], it may need minor adjustments (primarily, to use the settings centralized in `BlobStoreRepository`), otherwise the change should not break anything.

[1] https://github.com/opensearch-project/opensearch-oci-object-storage/

willyborankin commented:
> FYI, there is only one plugin (outside core) that implements repository for OCI [1], it may need minor adjustments (primarily, to use the settings centralized in `BlobStoreRepository`), otherwise the change should not break anything.
> 
> [1] https://github.com/opensearch-project/opensearch-oci-object-storage/

good point should I connect with them?

reta commented on server/src/main/java/org/opensearch/common/compress/ZstdCompressor.java:
```diff
@@ -0,0 +1,94 @@
+        out.write(HEADER);
+        return new ZstdOutputStreamNoFinalizer(new BufferedOutputStream(out, BUFFER_SIZE), RecyclingBufferPool.INSTANCE, LEVEL) {
+
+            // Due to https://bugs.openjdk.java.net/browse/JDK-8054565 we can't rely on the buffered output stream to only close once
```
Super minor: unless we aim to backport that to 1.x, we don't need this workaround anymore, the baseline for 2.x+ is JDK-11 where https://bugs.openjdk.java.net/browse/JDK-8054565 is fixed.

reta replied:
Plus, `ZstdOutputStreamNoFinalizer` has `isClosed` guards as well, we could safely drop this

willyborankin replied:
addressed

dblock commented:
> > FYI, there is only one plugin (outside core) that implements repository for OCI [1], it may need minor adjustments (primarily, to use the settings centralized in `BlobStoreRepository`), otherwise the change should not break anything.
> > [1] https://github.com/opensearch-project/opensearch-oci-object-storage/
> 
> good point should I connect with them?

Open an issue in that repo so we don't forget.

nknize reviewed (approved):
Haven't had a chance to revisit this, but it seems we have enough recent reviews and we can patch and/or revert hidden dragons. So I'll unblock.

nknize commented:
> @nknize mind to resolve your comments as well (may be something pops up)? (the merge is still blocked)

I would love to. But apparently you can't resolve outdated (missing) conversations. And the [alleged "workaround"](https://github.com/community/community/discussions/19206#discussioncomment-6030835) doesn't work on mobile view. I'm happy to merge as admin to override these pearly github gates.

nknize commented:
Thanks @willyborankin for your contribution and to everyone for your tenacity on this long running PR.

reta commented:
@willyborankin mind sending manual backport against `2.x`? thank you

willyborankin commented:
> @willyborankin mind sending manual backport against `2.x`? thank you

sure will do
